pyecasound/ecacontrol.py

```diff
@@ -1,291 +1,277 @@
-            elif I._type == 'li':
-                    I._resp[I._type]=int(r[1])
-            else:
-                    I._resp[I._type]=r[1]
+            self._resp[self._type] = r[1]
 
-        return I._resp[I._type]
+        return self._resp[self._type]
 
-    
-    def initialize(I):
+    def initialize(self):
         """Reserve resources"""
-                
-##        if _ecasound is not None:
-##            I.cleanup()             # exit previous ecasound session cleanly
-           
-        global _ecasound
 
         try:
-            ecasound_binary = os.environ['ECASOUND']
+            ecasound_binary = os.environ["ECASOUND"]
         except KeyError:
-            ecasound_binary = ''
-
-        if ecasound_binary == '':
-            ecasound_binary = 'ecasound'
-
-        p = subprocess.Popen(ecasound_binary + ' -c -d:256 2>/dev/null',
-                             shell=True, bufsize=0, stdin=subprocess.PIPE, stdout=subprocess.PIPE, close_fds=True)
-        _ecasound.append(p)
-        
-        I.eca=_ecasound[-1]
-        
-        lines=''
-        
-        lines=lines+I._readline()+'\n'
-
-        version=I._readline()
-            
-        s=str.find(version, 'ecasound v')
-        if float(version[s+10:s+13])>=2.2:
-            lines=lines+version+'\n'
+            ecasound_binary = ""
+
+        if ecasound_binary == "":
+            ecasound_binary = "ecasound"
+
+        self.eca = subprocess.Popen(
+            [ecasound_binary, "-c", "-E", "int-output-mode-wellformed"],
+            shell=False,
+            bufsize=0,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            close_fds=True,
+        )
+
+        lines = self._readline() + "\n"
+        version = self._readline()
+        s = version.find("ecasound v")
+        if float(version[s + 10: s + 13]) >= 2.2:
+            lines = lines + version + "\n"
         else:
-            raise RuntimeError('ecasound version 2.2 required!')
-        
-        lines=lines+I._readline()+'\n'
-        
-        if I.verbose:
+            raise RuntimeError("ecasound version 2.2+ required!")
+        lines = lines + self._readline() + "\n"
+
+        if self.verbose:
             print(lines)
             print(__doc__)
-            print('by', authors)
-            print('\n(to get rid of this message, pass zero to instance init)')
-            
-        I.command('int-output-mode-wellformed')
-        #I._read_eca()
-        #I.command('debug 256')
-        
-    def cleanup(I):
+            print("by", AUTHORS)
+            print("\n(to get rid of this message, pass zero to instance init)")
+
+        print(self._parse_response())
+
+        # self.command('debug 256')
+
+    def cleanup(self):
         """Free all reserved resources"""
-        
-        I.eca.stdin.write('quit\n'.encode())
 
-        os.kill(I.eca.pid, signal.SIGTERM)
-                
+        self.eca.stdin.write("quit\n")
```

Comment:
@marclava And here you need encode() as write() won't take a string object. 
Caught with "python3 ecacontrol.py".